[2.7] bpo-30759: Copy test_robotparser from master

vstinner opened the pull request:
Copy Lib/test/test_robotparser.py from master to 2.7 and adapt it for
Python 2.7:

* Replace urllib.robotparser with robotparser
* Adjust HTTPServer import
* Replace io.StringIO with StringIO.StringIO
* Remove tests on crawl_delay() and request_rate() since these
  methods were added to Python 3
* Remove subTest()
* Add test_main() which explicitly lists all test cases

vstinner committed:
bpo-30759: Copy test_robotparser from master

Copy Lib/test/test_robotparser.py from master to 2.7 and adapt it for
Python 2.7:

* Replace urllib.robotparser with robotparser
* Adjust HTTPServer import
* Replace io.StringIO with StringIO.StringIO
* Remove tests on crawl_delay() and request_rate() since these
  methods were added to Python 3
* Remove subTest()
* Add test_main() which explicitly lists all test cases

Patch based on the commit 4da0fd06ce49132d5c2e0525c4a31f807690a9ed
written by Berker Peksag.

Co-Authored-By: Berker Peksag <berker.peksag@gmail.com>

serhiy-storchaka commented on Lib/test/test_robotparser.py:
```diff
@@ -1,296 +1,237 @@
-from urllib2 import urlopen, HTTPError
-
-HAVE_HTTPS = True
+import io
```
Is it used?

vstinner replied:
removed

serhiy-storchaka commented on Lib/test/test_robotparser.py:
```diff
@@ -1,296 +1,237 @@
+import os
+import unittest
+import robotparser
+from collections import namedtuple
```
Is it used?

vstinner replied:
removed

serhiy-storchaka commented:
test_robotparser was rewritten in [bpo-25497](https://bugs.python.org/issue25497) (4da0fd06ce49132d5c2e0525c4a31f807690a9ed) by @berkerpeksag. Maybe try to cherry-pick 4da0fd06ce49132d5c2e0525c4a31f807690a9ed and compare results?

berkerpeksag commented on Lib/test/test_robotparser.py:
```diff
@@ -1,296 +1,237 @@
+        UseFirstUserAgentWildcardTest,
+        EmptyQueryStringTest,
+        DefaultEntryTest,
+        RobotHandler,
```
`RobotHandler` is a subclass of `BaseHTTPServer.BaseHTTPRequestHandler`, not `unittest.TestCase`.

vstinner replied:
Fixed. Funny that support.run_unittest() doesn't report such bug :-)

vstinner commented:
Oh, I don't know what happened: I did remove the 2 unused imports, but I see them in the first version of my PR? Well, it doesn't matter, I just removed them :-)

I fixed the 3 reports issues.

Serhiy: "test_robotparser was rewritten in bpo-25497 (4da0fd0) by @berkerpeksag. Maybe try to cherry-pick 4da0fd0 and compare results?"

Sorry, but I don't want to try to *cherry-pick* such large refactoring change from Python 3 to Python 2, the cherry-pick creates a file full of conflcts. IMHO it's simpler to start with a fresh *copy* of the file, and then adapt it for Python 2.

I credited @berkerpeksag in my commit message.

vstinner commented:
@serhiy-storchaka, @berkerpeksag: Does it look better now?

vstinner commented:
Ok, thank you for the review @berkerpeksag. Let's go!
